python/ray/serve/tests/test_proxy.py

```diff
@@ -202,101 +225,129 @@ def create_grpc_proxy(self):
             node_id=node_id,
             node_ip_address=node_ip_address,
             proxy_router_class=FakeProxyRouter,
-            controller_actor=FakeActorHandler("fake_controller_actor"),
         )
 
     @pytest.mark.asyncio
-    async def test_not_found(self):
-        """Test gRPCProxy set up the correct not found response."""
+    async def test_not_found_response(self):
+        """Test gRPCProxy returns the correct not found response."""
         grpc_proxy = self.create_grpc_proxy()
+        grpc_proxy.update_routes({})
 
-        # Application name isn't provided.
-        proxy_request = FakeProxyRequest()
-        proxy_request.app_name = ""
-        gen = grpc_proxy.not_found(proxy_request)
-        status = await gen.__anext__()
-        with pytest.raises(StopAsyncIteration):
-            await gen.__anext__()
+        # Application name not provided.
+        status, _ = await _consume_proxy_generator(
+            grpc_proxy.proxy_request(
+                FakeProxyRequest(request_type="grpc", app_name=""),
+            )
+        )
 
-        assert isinstance(status, ResponseStatus)
         assert status.code == grpc.StatusCode.NOT_FOUND
         assert "Application metadata not set" in status.message
         assert status.is_error is True
 
         # Application name is provided but wasn't found.
-        proxy_request.app_name = "foobar"
-        gen = grpc_proxy.not_found(proxy_request)
-        status = await gen.__anext__()
-        with pytest.raises(StopAsyncIteration):
-            await gen.__anext__()
+        status, _ = await _consume_proxy_generator(
+            grpc_proxy.proxy_request(
+                FakeProxyRequest(request_type="grpc", app_name="foobar"),
+            )
+        )
 
-        assert isinstance(status, ResponseStatus)
         assert status.code == grpc.StatusCode.NOT_FOUND
         assert "Application 'foobar' not found" in status.message
         assert status.is_error is True
 
     @pytest.mark.asyncio
-    async def test_draining_response(self):
-        """Test gRPCProxy set up the correct draining response."""
+    @pytest.mark.parametrize("is_draining", [False, True])
+    @pytest.mark.parametrize("routes_updated", [False, True])
+    async def test_routes_response(self, is_draining: bool, routes_updated: bool):
```

Comment:
@GeneDer please have a look at how I've updated these methods to test the behavior in a more end-to-end fashion. We're no longer testing just the specify `xyz_response` methods, but instead the full path in the proxy that calls them.

Reply:
LGTM, we also have integration tests in `test_grpc.py` I believe, but  those are good too!